### ISSUE TITLE
vars can't be used with tls or root option

### ISSUE THREAD

**Issue by coolaj86** [mholt/caddy#125]
This probably won't become possible until the API work is done anyway, but I was hoping it would:

```
0.0.0.0:8443 {
  tls /srv/vhosts/{hostname}/certs/public.bundle.pem /srv/vhosts/{hostname}/certs/private.key.pem
  root /srv/vhosts/{hostname}/public
}
```

**Comment by abiosoft** [comment]
See my response to the other similar issue. https://github.com/mholt/caddy/issues/124#issuecomment-110648127

**Comment by mholt** [comment]
As cool as that would be, using request-dependent variables in the tls and root directives is nearly impossible, I think. Both of those directives are compiled before the server even starts. Even if it did work, if two requests come in at the same time (it could happen), there would be lock contention over making the change to the server for handling that one request.

I can see the value of what you are suggesting here, but I'm not sure it's technically feasible...

**Comment by coolaj86** [comment]
I'll have to become more familiar with the code to understand what you mean.

If you're able to handle the HUP case and add API stuff, this will probably fit in with those changes.

**Comment by tw4452852** [comment]
Maybe we could make the CaddyFile a template, so we could auto fill something, like host name or port if any, like this:

```
localhost:8443 {
  tls /srv/vhosts/{hostname}_{port}/certs/public.bundle.pem 
  root /srv/vhosts/{hostname}_{port}/public
}
```

will be rendered into:

```
localhost:8443 {
  tls /srv/vhosts/localhost_8443/certs/public.bundle.pem 
  root /srv/vhosts/localhost_8443/public
}
```

Anyway this is just my thought, :smile:

**Comment by mholt** [comment]
@tw4452852 Interesting idea, I will consider it, to see if there's a good way to make this happen. (By the way, I just now realized that your avatar is Totorow. I watched My Neighbor Totorow with my nieces and nephew the other day and really liked the story! One of my favorites.)

**Comment by mholt** [comment]
This is closely related to #167 - fixing one should fix the other (at least as far as "root" goes - tls is another matter which I hope to address soon).

**Comment by mholt** [comment]
So which site root do we give to directives that run startup functions (like `markdown` doing static site generation, or `git` pulling updates) if we don't know the site root until request-time?

**Comment by mholt** [comment]
Related discussion is in #619. If proposal 2 is implemented there, it eliminates the need for variable root path, and eliminates the problem about knowing what the site root is at startup.

Variables in the paths for tls can also be eliminated with updates available in version 0.8.2 by specifying a directory to load certificates from, which will automatically be keyed to hostname anyway.

Phew :sweat_smile:
